### ISSUE TITLE
Creating a IotHubResourceEventHubConsumerGroup fails on second `pulumi up`

### ISSUE THREAD

**Issue by ChristianEder** [pulumi/pulumi-azure-native#876]
Defining IoT Hub consumer groups using the IotHubResourceEventHubConsumerGroup resource fails on subsequent calls to `pulumi up`.

## Expected behavior

As defined in the docs, creating an IotHubResourceEventHubConsumerGroup should create a new consumer group on the first run of `pulumi up` and not fail on subsequent `pulumi up` calls.

## Current behavior

An IoT Hub with a custom consumer group in addition to the standard "$Default" consumer group is created when running the following stack on the first run of `pulumi up` (up until here, the behaviour is as expected):

```csharp
class MyStack: Stack
{
    public MyStack()
    {
        var resourceGroup = new ResourceGroup("my-rg", new ResourceGroupArgs { ResourceGroupName = "my-rg" });
        var iotHub = new IotHubResource("my-iot-hub", new IotHubResourceArgs
        {
            ResourceName = "my-iot-hub",
            ResourceGroupName = resourceGroup.Name,
            Sku = new IotHubSkuInfoArgs
            {
                Name = IotHubSku.B1,
                Capacity = 1
            }
        });
        var consumerGroup = new IotHubResourceEventHubConsumerGroup("myconsumergroup", new IotHubResourceEventHubConsumerGroupArgs
        {
            ResourceGroupName = resourceGroup.Name,
            ResourceName = iotHub.Name,
            EventHubEndpointName = "events",
            Name = "myconsumergroup",
            Properties = new EventHubConsumerGroupNameArgs { Name = "myconsumergroup" }
        });
    }
}
```

On the second call to `pulumi up` (no matter if any changes to the stack where made or not, so it also fails during `pulumi preview`), the following error message is received:

```
Previewing update (dev):
     Type                 Name                                  Plan     Info
     pulumi:pulumi:Stack  ******           1 error

Diagnostics:
  pulumi:pulumi:Stack (******):
    error: Running program '******.dll' failed with an unhandled exception:
    System.InvalidOperationException: Expected System.String but got System.Collections.Immutable.ImmutableDictionary`2[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]] deserializing Pulumi.AzureNative.Devices.IotHubResourceEventHubConsumerGroup.properties
       at Pulumi.Serialization.Converter.ConvertObject(String context, Object val, Type targetType)
       at Pulumi.Serialization.Converter.ConvertValue(String context, Value value, Type targetType, ImmutableHashSet`1 resources)
       at Pulumi.Deployment.CompleteResourceAsync(Resource resource, Boolean remote, Func`2 newDependency, ResourceArgs args, ResourceOptions options, ImmutableDictionary`2 completionSources)
       at Pulumi.Deployment.Runner.<>c__DisplayClass9_0.<<WhileRunningAsync>g__HandleCompletion|0>d.MoveNext()
    --- End of stack trace from previous location ---
       at Pulumi.Deployment.Runner.WhileRunningAsync()
```

## Steps to reproduce

see "Current behaviour"

## Context (Environment)

I am trying to create an Azure IoT Hub resource with two consumer groups: the "$Default" plus an additional one.

I am using the .NET SDK "Pulumi.AzureNative" in version 1.5.0 from a .NET 5 application using the pulumi CLI version 3.1.0 from a Windows 10 PC.

**Comment by ChristianEder** [comment]
I tried updating the "Pulumi.AzureNative" nuget to the latest version "1.8.0" and the pulumi CLI to the latest version "3.3.1" and re-ran `pulumi preview` - but still get the same error message.

**Comment by mikhailshilkov** [comment]
Sorry for the trouble @ChristianEder 

What happens if you run `pulumi refresh`? What is the diff that you see? Does it work? If yes, does the update work after it?

**Comment by ChristianEder** [comment]
Besides a whole lot of other diffs, it shows:

`├─ azure-native:devices:IotHubResourceEventHubConsumerGroup  myconsumergroup                                 update     [diff: ~properties]`

Here the full diff:

```
Previewing refresh (dev):
     Type                                                         Name                                  Plan       Info
     pulumi:pulumi:Stack                                          ******
     ├─ azure-native:resources:ResourceGroup                      ******
     ├─ azure-native:servicebus:QueueAuthorizationRule            ******
     ├─ azure-native:servicebus:Namespace                         ******
 ~   ├─ azure-native:servicebus:Queue                             ******                      update     [diff: ~accessedAt,countDetails,createdAt,messageCount,sizeInBytes,updatedAt]
 ~   ├─ azure-native:storage:BlobContainer                        ******                             update     [diff: +defaultEncryptionScope,denyEncryptionScopeOverride,etag,lastModifiedTime,leaseState,leaseS     
     ├─ azure-native:servicebus:QueueAuthorizationRule            ******
     ├─ azure-native:insights:Component                          ******
 ~   ├─ azure-native:storage:BlobContainer                        ******                               update     [diff: +defaultEncryptionScope,denyEncryptionScopeOverride,etag,lastModifiedTime,leaseState,leaseS     
 ~   ├─ azure-native:storage:BlobContainer                       ******                           update     [diff: +defaultEncryptionScope,denyEncryptionScopeOverride,etag,lastModifiedTime,leaseState,leaseS     
     ├─ azure-native:storage:Blob                                 ******.zip
 ~   ├─ azure-native:devices:IotHubResourceEventHubConsumerGroup  myconsumergroup                                 update     [diff: ~properties]
     ├─ azure-native:storage:StorageAccount                       ******
 ~   ├─ azure-native:web:AppServicePlan                           ******                       update     [diff: ~numberOfSites]
 ~   ├─ azure-native:web:WebApp                                   ******                       update     [diff: ~lastModifiedTimeUtc]
     └─ azure-native:devices:IotHubResource                       ******

Resources:
    ~ 7 to update
    9 unchanged
```
After successfully running the refresh, `pulumi preview` also works again, but wants to do some changes:

```
Previewing update (dev):
     Type                                   Name                                  Plan       Info
     pulumi:pulumi:Stack                    zuehlke-iot-infrastructure-azure-dev
 ~   ├─ azure-native:storage:BlobContainer  *****                               update     [diff: -defaultEncryptionScope,denyEncryptionScopeOverride]
 ~   ├─ azure-native:storage:BlobContainer  *****                             update     [diff: -defaultEncryptionScope,denyEncryptionScopeOverride]
 ~   └─ azure-native:storage:BlobContainer  *****                           update     [diff: -defaultEncryptionScope,denyEncryptionScopeOverride]
```

A `pulumi up` run after that also succeeds. And yet another `pulumi preview` and `pulumi up` run after that also succeeds, with no changes applied. Seems to be back into a working state now. 

Does this mean a `pulumi refresh` is required after the first `pulumi up` for a new stack?

**Comment by mikhailshilkov** [comment]
I can reproduce the issue. Here is the response that Azure returns to the PUT operation:

```
{"properties":{"created":"Wed, 26 May 2021 12:01:54 GMT","properties":{"name":"myconsumergroup"}},"id":"/subscriptions/<sub>/resourceGroups/my-rg0fe28958/providers/Microsoft.Devices/IotHubs/my-iot-hub106dfdde/eventHubEndpoints/events/ConsumerGroups/myconsumergroup","name":"myconsumergroup","type":"Microsoft.Devices/IotHubs/EventHubEndpoints/ConsumerGroups","etag":null}
```

Note how there's a `properties` object inside another `properties` object. I think it's a bug of the Azure API that is fixed by the GET endpoint (see below).

The default API version `2020-08-31` that we are using [describes](https://github.com/Azure/azure-rest-api-specs/blob/3ce916b44254c9fde18d4fb01ae021e8cd2bf689/specification/iothub/resource-manager/Microsoft.Devices/stable/2020-08-31/iothub.json#L3620-L3625) this property as a dictionary of strings, which isn't correct in this case.

The later API version has actually [changed](https://github.com/Azure/azure-rest-api-specs/blob/3ce916b44254c9fde18d4fb01ae021e8cd2bf689/specification/iothub/resource-manager/Microsoft.Devices/stable/2021-03-31/iothub.json#L3649-L3652) this definition to be a dictionary of objects, which would prevent the error.

Your options for now:

1. Run a refresh after creating `IotHubResourceEventHubConsumerGroup`. I'm actually not sure why the `properties` collection isn't broken anymore, but `GET` doesn't return non-string values, so the state is fixed.
2. Use the `V20210331` namespace instead of the default one in your program.

Potential improvements for the provider:

1. Promote the `IotHubResourceEventHubConsumerGroup` to the `2021-03-31` version for the top-level resource.
2. Add an extra validation for the dictionaries that we receive from Azure and ignore invalid values with a warning.

How does that sound?

**Comment by ChristianEder** [comment]
Sounds great, thanks for identifying and solving the issue so fast, awesome!

**Comment by mikhailshilkov** [comment]
The underlying issue in the .NET SDK was fixed in pulumi 3.17.0 and shipped in azure-native 1.45.0. Even if data types don't match, there will be no exception thrown by Pulumi. I'll go ahead and close this.
